serde-generate/runtime/typescript/serde/binaryDeserializer.ts

```diff
@@ -0,0 +1,129 @@
+import { Deserializer } from './deserializer';
+
+export abstract class BinaryDeserializer implements Deserializer {
+  private static readonly BIG_32 = BigInt(32);
+  private static readonly BIG_64 = BigInt(64);
+  private static readonly textDecoder: TextDecoder = new TextDecoder();
+  public buffer: ArrayBuffer;
+  public offset: number;
+
+  protected constructor(data: Uint8Array) {
+    this.buffer = data;
+    this.offset = 0;
+  }
+
+  private read(length: number): Uint8Array {
+    const bytes = new Uint8Array(
+        this.buffer.slice(this.offset, this.offset + length)
+    );
+    this.offset += length;
+    return bytes;
+  }
+
+  abstract deserializeLen(): number;
+
+  abstract deserializeVariantIndex(): number;
+
+  abstract checkThatKeySlicesAreIncreasing(
+      key1: [number, number],
+      key2: [number, number]
+  ): void;
+
+  public deserializeStr(): string {
+    const value = this.deserializeBytes();
+    return BinaryDeserializer.textDecoder.decode(value);
+  }
+
+  public deserializeBytes(): Uint8Array {
+    const len = this.deserializeLen();
+    if (len < 0) {
+      throw new Error("Length of a bytes array can't be negative");
+    }
+    return this.read(len);
+  }
+
+  public deserializeBool(): boolean {
+    const bool = this.read(1)[0];
+    return bool == 1;
+  }
+
+  public deserializeUnit(): undefined {
+    return;
+  }
+
+  public deserializeU8(): number {
+    return new DataView(this.read(1)).getUint8(0);
+  }
+
+  public deserializeU16(): number {
+    return new DataView(this.read(2)).getUint16(0, true);
+  }
+
+  public deserializeU32(): number {
+    return new DataView(this.read(4)).getUint32(0, true);
+  }
+
+  public deserializeU64(): BigInt {
+    const left = this.deserializeU32();
```

Comment:
I found it very confusing, as when deserializing a little-endian integer, what would be the low bits? the first to be read are the lowest, but when reading it, it's at the high side of the read. I used left and right as an absolute marker. If you think it's better to use high and low, I could change it :-)

Reply:
I would argue that "low" and "high" are appropriate conventions for a little endian encoding because "low" conveys both the idea of least-significant (low-exponent) bit and the idea of a low address in memory.

Conversely, I'm a bit confused why the least significant part of an integer would be called "left".

Reply:
changed to low and high